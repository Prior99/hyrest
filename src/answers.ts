import * as HTTP from "http-status-codes";

export interface Answer<T> {
    readonly message?: string;
    readonly data?: T;
}

export interface WrappedAnswer<T> {
    readonly statusCode: number;
    readonly result: Answer<T>;
}

export interface Wrapper<T> {
    readonly body?: T;
    readonly message?: string;

}

function isWrapper(wrapper: any): wrapper is Wrapper<any> {
    if (typeof wrapper !== "object") {
        return false;
    }
    return Object.keys(wrapper).every(key => ["message", "body"].includes(key));
}

function answer(statusCode: number, ...args: any[]): any {
    if (args.length === 0) {
        return {
            statusCode,
        };
    }
    if (args.length === 1) {
        if (isWrapper(args[0])) {
            const { message, data } = args[0];
            return {
                statusCode,
                result: {
                    message,
                    data,
                },
            };
        }
        if (typeof args[0] === "string") {
            return {
                statusCode,
                result: {
                    message: args[0],
                },
            };
        }
        return {
            statusCode,
            result: {
                data: args[0],
            },
        };
    }
    return {
        statusCode,
        result: {
            data: args[0],
            message: args[1],
        },
    };
}

export function accepted<T>(wrapper: Wrapper<T>): T;
export function accepted<T>(body?: T): T;
export function accepted<T>(message?: string): T;
export function accepted<T>(body?: T, message?: string): T;
export function accepted<T>(...args: any[]): T {
    return answer(HTTP.ACCEPTED, ...args) as T;
}

export function badGateway<T>(wrapper: Wrapper<T>): T;
export function badGateway<T>(body?: T): T;
export function badGateway<T>(message?: string): T;
export function badGateway<T>(body?: T, message?: string): T;
export function badGateway<T>(...args: any[]): T {
    return answer(HTTP.BAD_GATEWAY, ...args) as T;
}

export function badRequest<T>(wrapper: Wrapper<T>): T;
export function badRequest<T>(body?: T): T;
export function badRequest<T>(message?: string): T;
export function badRequest<T>(body?: T, message?: string): T;
export function badRequest<T>(...args: any[]): T {
    return answer(HTTP.BAD_REQUEST, ...args) as T;
}

export function conflict<T>(wrapper: Wrapper<T>): T;
export function conflict<T>(body?: T): T;
export function conflict<T>(message?: string): T;
export function conflict<T>(body?: T, message?: string): T;
export function conflict<T>(...args: any[]): T {
    return answer(HTTP.CONFLICT, ...args) as T;
}

export function httpContinue<T>(wrapper: Wrapper<T>): T;
export function httpContinue<T>(body?: T): T;
export function httpContinue<T>(message?: string): T;
export function httpContinue<T>(body?: T, message?: string): T;
export function httpContinue<T>(...args: any[]): T {
    return answer(HTTP.CONTINUE, ...args) as T;
}

export function created<T>(wrapper: Wrapper<T>): T;
export function created<T>(body?: T): T;
export function created<T>(message?: string): T;
export function created<T>(body?: T, message?: string): T;
export function created<T>(...args: any[]): T {
    return answer(HTTP.CREATED, ...args) as T;
}

export function expectationFailed<T>(wrapper: Wrapper<T>): T;
export function expectationFailed<T>(body?: T): T;
export function expectationFailed<T>(message?: string): T;
export function expectationFailed<T>(body?: T, message?: string): T;
export function expectationFailed<T>(...args: any[]): T {
    return answer(HTTP.EXPECTATION_FAILED, ...args) as T;
}

export function failedDependency<T>(wrapper: Wrapper<T>): T;
export function failedDependency<T>(body?: T): T;
export function failedDependency<T>(message?: string): T;
export function failedDependency<T>(body?: T, message?: string): T;
export function failedDependency<T>(...args: any[]): T {
    return answer(HTTP.FAILED_DEPENDENCY, ...args) as T;
}

export function forbidden<T>(wrapper: Wrapper<T>): T;
export function forbidden<T>(body?: T): T;
export function forbidden<T>(message?: string): T;
export function forbidden<T>(body?: T, message?: string): T;
export function forbidden<T>(...args: any[]): T {
    return answer(HTTP.FORBIDDEN, ...args) as T;
}

export function gatewayTimeout<T>(wrapper: Wrapper<T>): T;
export function gatewayTimeout<T>(body?: T): T;
export function gatewayTimeout<T>(message?: string): T;
export function gatewayTimeout<T>(body?: T, message?: string): T;
export function gatewayTimeout<T>(...args: any[]): T {
    return answer(HTTP.GATEWAY_TIMEOUT, ...args) as T;
}

export function gone<T>(wrapper: Wrapper<T>): T;
export function gone<T>(body?: T): T;
export function gone<T>(message?: string): T;
export function gone<T>(body?: T, message?: string): T;
export function gone<T>(...args: any[]): T {
    return answer(HTTP.GONE, ...args) as T;
}

export function httpVersionNotSupported<T>(wrapper: Wrapper<T>): T;
export function httpVersionNotSupported<T>(body?: T): T;
export function httpVersionNotSupported<T>(message?: string): T;
export function httpVersionNotSupported<T>(body?: T, message?: string): T;
export function httpVersionNotSupported<T>(...args: any[]): T {
    return answer(HTTP.HTTP_VERSION_NOT_SUPPORTED, ...args) as T;
}

export function insufficientSpaceOnResource<T>(wrapper: Wrapper<T>): T;
export function insufficientSpaceOnResource<T>(body?: T): T;
export function insufficientSpaceOnResource<T>(message?: string): T;
export function insufficientSpaceOnResource<T>(body?: T, message?: string): T;
export function insufficientSpaceOnResource<T>(...args: any[]): T {
    return answer(HTTP.INSUFFICIENT_SPACE_ON_RESOURCE, ...args) as T;
}

export function insufficientStorage<T>(wrapper: Wrapper<T>): T;
export function insufficientStorage<T>(body?: T): T;
export function insufficientStorage<T>(message?: string): T;
export function insufficientStorage<T>(body?: T, message?: string): T;
export function insufficientStorage<T>(...args: any[]): T {
    return answer(HTTP.INSUFFICIENT_STORAGE, ...args) as T;
}

export function internalServerError<T>(wrapper: Wrapper<T>): T;
export function internalServerError<T>(body?: T): T;
export function internalServerError<T>(message?: string): T;
export function internalServerError<T>(body?: T, message?: string): T;
export function internalServerError<T>(...args: any[]): T {
    return answer(HTTP.INTERNAL_SERVER_ERROR, ...args) as T;
}

export function lengthRequired<T>(wrapper: Wrapper<T>): T;
export function lengthRequired<T>(body?: T): T;
export function lengthRequired<T>(message?: string): T;
export function lengthRequired<T>(body?: T, message?: string): T;
export function lengthRequired<T>(...args: any[]): T {
    return answer(HTTP.LENGTH_REQUIRED, ...args) as T;
}

export function locked<T>(wrapper: Wrapper<T>): T;
export function locked<T>(body?: T): T;
export function locked<T>(message?: string): T;
export function locked<T>(body?: T, message?: string): T;
export function locked<T>(...args: any[]): T {
    return answer(HTTP.LOCKED, ...args) as T;
}

export function methodFailure<T>(wrapper: Wrapper<T>): T;
export function methodFailure<T>(body?: T): T;
export function methodFailure<T>(message?: string): T;
export function methodFailure<T>(body?: T, message?: string): T;
export function methodFailure<T>(...args: any[]): T {
    return answer(HTTP.METHOD_FAILURE, ...args) as T;
}

export function methodNotAllowed<T>(wrapper: Wrapper<T>): T;
export function methodNotAllowed<T>(body?: T): T;
export function methodNotAllowed<T>(message?: string): T;
export function methodNotAllowed<T>(body?: T, message?: string): T;
export function methodNotAllowed<T>(...args: any[]): T {
    return answer(HTTP.METHOD_NOT_ALLOWED, ...args) as T;
}

export function movedPermanently<T>(wrapper: Wrapper<T>): T;
export function movedPermanently<T>(body?: T): T;
export function movedPermanently<T>(message?: string): T;
export function movedPermanently<T>(body?: T, message?: string): T;
export function movedPermanently<T>(...args: any[]): T {
    return answer(HTTP.MOVED_PERMANENTLY, ...args) as T;
}

export function movedTemporarily<T>(wrapper: Wrapper<T>): T;
export function movedTemporarily<T>(body?: T): T;
export function movedTemporarily<T>(message?: string): T;
export function movedTemporarily<T>(body?: T, message?: string): T;
export function movedTemporarily<T>(...args: any[]): T {
    return answer(HTTP.MOVED_TEMPORARILY, ...args) as T;
}

export function multiStatus<T>(wrapper: Wrapper<T>): T;
export function multiStatus<T>(body?: T): T;
export function multiStatus<T>(message?: string): T;
export function multiStatus<T>(body?: T, message?: string): T;
export function multiStatus<T>(...args: any[]): T {
    return answer(HTTP.MULTI_STATUS, ...args) as T;
}

export function multipleChoices<T>(wrapper: Wrapper<T>): T;
export function multipleChoices<T>(body?: T): T;
export function multipleChoices<T>(message?: string): T;
export function multipleChoices<T>(body?: T, message?: string): T;
export function multipleChoices<T>(...args: any[]): T {
    return answer(HTTP.MULTIPLE_CHOICES, ...args) as T;
}

export function networkAuthenticationRequired<T>(wrapper: Wrapper<T>): T;
export function networkAuthenticationRequired<T>(body?: T): T;
export function networkAuthenticationRequired<T>(message?: string): T;
export function networkAuthenticationRequired<T>(body?: T, message?: string): T;
export function networkAuthenticationRequired<T>(...args: any[]): T {
    return answer(HTTP.NETWORK_AUTHENTICATION_REQUIRED, ...args) as T;
}

export function noContent<T>(wrapper: Wrapper<T>): T;
export function noContent<T>(body?: T): T;
export function noContent<T>(message?: string): T;
export function noContent<T>(body?: T, message?: string): T;
export function noContent<T>(...args: any[]): T {
    return answer(HTTP.NO_CONTENT, ...args) as T;
}

export function nonAuthoritativeInformation<T>(wrapper: Wrapper<T>): T;
export function nonAuthoritativeInformation<T>(body?: T): T;
export function nonAuthoritativeInformation<T>(message?: string): T;
export function nonAuthoritativeInformation<T>(body?: T, message?: string): T;
export function nonAuthoritativeInformation<T>(...args: any[]): T {
    return answer(HTTP.NON_AUTHORITATIVE_INFORMATION, ...args) as T;
}

export function notAcceptable<T>(wrapper: Wrapper<T>): T;
export function notAcceptable<T>(body?: T): T;
export function notAcceptable<T>(message?: string): T;
export function notAcceptable<T>(body?: T, message?: string): T;
export function notAcceptable<T>(...args: any[]): T {
    return answer(HTTP.NOT_ACCEPTABLE, ...args) as T;
}

export function notFound<T>(wrapper: Wrapper<T>): T;
export function notFound<T>(body?: T): T;
export function notFound<T>(message?: string): T;
export function notFound<T>(body?: T, message?: string): T;
export function notFound<T>(...args: any[]): T {
    return answer(HTTP.NOT_FOUND, ...args) as T;
}

export function notImplemented<T>(wrapper: Wrapper<T>): T;
export function notImplemented<T>(body?: T): T;
export function notImplemented<T>(message?: string): T;
export function notImplemented<T>(body?: T, message?: string): T;
export function notImplemented<T>(...args: any[]): T {
    return answer(HTTP.NOT_IMPLEMENTED, ...args) as T;
}

export function notModified<T>(wrapper: Wrapper<T>): T;
export function notModified<T>(body?: T): T;
export function notModified<T>(message?: string): T;
export function notModified<T>(body?: T, message?: string): T;
export function notModified<T>(...args: any[]): T {
    return answer(HTTP.NOT_MODIFIED, ...args) as T;
}

export function ok<T>(wrapper: Wrapper<T>): T;
export function ok<T>(body?: T): T;
export function ok<T>(message?: string): T;
export function ok<T>(body?: T, message?: string): T;
export function ok<T>(...args: any[]): T {
    return answer(HTTP.OK, ...args) as T;
}

export function partialContent<T>(wrapper: Wrapper<T>): T;
export function partialContent<T>(body?: T): T;
export function partialContent<T>(message?: string): T;
export function partialContent<T>(body?: T, message?: string): T;
export function partialContent<T>(...args: any[]): T {
    return answer(HTTP.PARTIAL_CONTENT, ...args) as T;
}

export function paymentRequired<T>(wrapper: Wrapper<T>): T;
export function paymentRequired<T>(body?: T): T;
export function paymentRequired<T>(message?: string): T;
export function paymentRequired<T>(body?: T, message?: string): T;
export function paymentRequired<T>(...args: any[]): T {
    return answer(HTTP.PAYMENT_REQUIRED, ...args) as T;
}

export function preconditionFailed<T>(wrapper: Wrapper<T>): T;
export function preconditionFailed<T>(body?: T): T;
export function preconditionFailed<T>(message?: string): T;
export function preconditionFailed<T>(body?: T, message?: string): T;
export function preconditionFailed<T>(...args: any[]): T {
    return answer(HTTP.PRECONDITION_FAILED, ...args) as T;
}

export function preconditionRequired<T>(wrapper: Wrapper<T>): T;
export function preconditionRequired<T>(body?: T): T;
export function preconditionRequired<T>(message?: string): T;
export function preconditionRequired<T>(body?: T, message?: string): T;
export function preconditionRequired<T>(...args: any[]): T {
    return answer(HTTP.PRECONDITION_REQUIRED, ...args) as T;
}

export function processing<T>(wrapper: Wrapper<T>): T;
export function processing<T>(body?: T): T;
export function processing<T>(message?: string): T;
export function processing<T>(body?: T, message?: string): T;
export function processing<T>(...args: any[]): T {
    return answer(HTTP.PROCESSING, ...args) as T;
}

export function proxyAuthenticationRequired<T>(wrapper: Wrapper<T>): T;
export function proxyAuthenticationRequired<T>(body?: T): T;
export function proxyAuthenticationRequired<T>(message?: string): T;
export function proxyAuthenticationRequired<T>(body?: T, message?: string): T;
export function proxyAuthenticationRequired<T>(...args: any[]): T {
    return answer(HTTP.PROXY_AUTHENTICATION_REQUIRED, ...args) as T;
}

export function requestHeaderFieldsTooLarge<T>(wrapper: Wrapper<T>): T;
export function requestHeaderFieldsTooLarge<T>(body?: T): T;
export function requestHeaderFieldsTooLarge<T>(message?: string): T;
export function requestHeaderFieldsTooLarge<T>(body?: T, message?: string): T;
export function requestHeaderFieldsTooLarge<T>(...args: any[]): T {
    return answer(HTTP.REQUEST_HEADER_FIELDS_TOO_LARGE, ...args) as T;
}

export function requestTimeout<T>(wrapper: Wrapper<T>): T;
export function requestTimeout<T>(body?: T): T;
export function requestTimeout<T>(message?: string): T;
export function requestTimeout<T>(body?: T, message?: string): T;
export function requestTimeout<T>(...args: any[]): T {
    return answer(HTTP.REQUEST_TIMEOUT, ...args) as T;
}

export function requestTooLong<T>(wrapper: Wrapper<T>): T;
export function requestTooLong<T>(body?: T): T;
export function requestTooLong<T>(message?: string): T;
export function requestTooLong<T>(body?: T, message?: string): T;
export function requestTooLong<T>(...args: any[]): T {
    return answer(HTTP.REQUEST_TOO_LONG, ...args) as T;
}

export function requestUriTooLong<T>(wrapper: Wrapper<T>): T;
export function requestUriTooLong<T>(body?: T): T;
export function requestUriTooLong<T>(message?: string): T;
export function requestUriTooLong<T>(body?: T, message?: string): T;
export function requestUriTooLong<T>(...args: any[]): T {
    return answer(HTTP.REQUEST_URI_TOO_LONG, ...args) as T;
}

export function requestedRangeNotSatisfiable<T>(wrapper: Wrapper<T>): T;
export function requestedRangeNotSatisfiable<T>(body?: T): T;
export function requestedRangeNotSatisfiable<T>(message?: string): T;
export function requestedRangeNotSatisfiable<T>(body?: T, message?: string): T;
export function requestedRangeNotSatisfiable<T>(...args: any[]): T {
    return answer(HTTP.REQUESTED_RANGE_NOT_SATISFIABLE, ...args) as T;
}

export function resetContent<T>(wrapper: Wrapper<T>): T;
export function resetContent<T>(body?: T): T;
export function resetContent<T>(message?: string): T;
export function resetContent<T>(body?: T, message?: string): T;
export function resetContent<T>(...args: any[]): T {
    return answer(HTTP.RESET_CONTENT, ...args) as T;
}

export function seeOther<T>(wrapper: Wrapper<T>): T;
export function seeOther<T>(body?: T): T;
export function seeOther<T>(message?: string): T;
export function seeOther<T>(body?: T, message?: string): T;
export function seeOther<T>(...args: any[]): T {
    return answer(HTTP.SEE_OTHER, ...args) as T;
}

export function serviceUnavailable<T>(wrapper: Wrapper<T>): T;
export function serviceUnavailable<T>(body?: T): T;
export function serviceUnavailable<T>(message?: string): T;
export function serviceUnavailable<T>(body?: T, message?: string): T;
export function serviceUnavailable<T>(...args: any[]): T {
    return answer(HTTP.SERVICE_UNAVAILABLE, ...args) as T;
}

export function switchingProtocols<T>(wrapper: Wrapper<T>): T;
export function switchingProtocols<T>(body?: T): T;
export function switchingProtocols<T>(message?: string): T;
export function switchingProtocols<T>(body?: T, message?: string): T;
export function switchingProtocols<T>(...args: any[]): T {
    return answer(HTTP.SWITCHING_PROTOCOLS, ...args) as T;
}

export function temporaryRedirect<T>(wrapper: Wrapper<T>): T;
export function temporaryRedirect<T>(body?: T): T;
export function temporaryRedirect<T>(message?: string): T;
export function temporaryRedirect<T>(body?: T, message?: string): T;
export function temporaryRedirect<T>(...args: any[]): T {
    return answer(HTTP.TEMPORARY_REDIRECT, ...args) as T;
}

export function tooManyRequests<T>(wrapper: Wrapper<T>): T;
export function tooManyRequests<T>(body?: T): T;
export function tooManyRequests<T>(message?: string): T;
export function tooManyRequests<T>(body?: T, message?: string): T;
export function tooManyRequests<T>(...args: any[]): T {
    return answer(HTTP.TOO_MANY_REQUESTS, ...args) as T;
}

export function unauthorized<T>(wrapper: Wrapper<T>): T;
export function unauthorized<T>(body?: T): T;
export function unauthorized<T>(message?: string): T;
export function unauthorized<T>(body?: T, message?: string): T;
export function unauthorized<T>(...args: any[]): T {
    return answer(HTTP.UNAUTHORIZED, ...args) as T;
}

export function unprocessableEntity<T>(wrapper: Wrapper<T>): T;
export function unprocessableEntity<T>(body?: T): T;
export function unprocessableEntity<T>(message?: string): T;
export function unprocessableEntity<T>(body?: T, message?: string): T;
export function unprocessableEntity<T>(...args: any[]): T {
    return answer(HTTP.UNPROCESSABLE_ENTITY, ...args) as T;
}

export function unsupportedMediaType<T>(wrapper: Wrapper<T>): T;
export function unsupportedMediaType<T>(body?: T): T;
export function unsupportedMediaType<T>(message?: string): T;
export function unsupportedMediaType<T>(body?: T, message?: string): T;
export function unsupportedMediaType<T>(...args: any[]): T {
    return answer(HTTP.UNSUPPORTED_MEDIA_TYPE, ...args) as T;
}

export function useProxy<T>(wrapper: Wrapper<T>): T;
export function useProxy<T>(body?: T): T;
export function useProxy<T>(message?: string): T;
export function useProxy<T>(body?: T, message?: string): T;
export function useProxy<T>(...args: any[]): T {
    return answer(HTTP.USE_PROXY, ...args) as T;
}
