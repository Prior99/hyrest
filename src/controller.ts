import { Constructable } from "./types";
import "isomorphic-fetch";
import "reflect-metadata";

import { Route } from "./route";
import { Params, ApiError, HTTPMethod } from "./types";
import { compile } from "path-to-regexp";
import { isBrowser } from "./is-browser";
import { populate } from "./scope";

export enum ControllerMode {
    SERVER = "server",
    CLIENT = "client",
}

export function getDefaultControllerMode() {
    return isBrowser() ? ControllerMode.CLIENT : ControllerMode.SERVER;
}

function buildQueryString(query: Params) {
    const queryString = Object.keys(query).reduce((result, key) => {
        const unencoded = query[key];
        if (typeof unencoded === "undefined" || unencoded === null) {
            return result;
        }
        const value = encodeURIComponent(unencoded);
        return `${result}&${key}=${value}`;
    }, "");
    // Remove first `&`.
    return queryString.substr(1, queryString.length);
}

/**
 * Assembles a full Url from an express url template, url parameters, as base Url and a query parameters.
 *
 * @param urlParameters The parameters fed to the express url template specified in `subUrl`.
 * @param query The query parameters for the url.
 * @param baseUrl The url to which the template url in `subUrl` is relative to.
 * @param subUrl An url template as used in express.
 *
 * @return The assembled Url.
 */
export function buildUrl(urlParameters: Params, query: Params, baseUrl: string, subUrl: string) {
    const queryString = buildQueryString(query);
    // The url parameters need to be explicitly converted to string as they will be transported via
    // the url string.
    const stringifiedParameters = Object.keys(urlParameters).reduce((result, key) => {
        result[key] = `${urlParameters[key]}`;
        return result;
    }, {} as any);

    // Compile the route's pattern (This is the reverse of what express does with url patterns) and
    // feed it the url parameters which had been converted to string.
    const routeString = compile(subUrl)(stringifiedParameters);

    // Assemble the full url.
    return `${baseUrl}${routeString}?${queryString}`;
}

/**
 * A handler which will be called whenever an Api error is encountered.
 *
 * @param error The Api error that was encountered.
 */
export type ErrorHandler = (error: ApiError) => void;

/**
 * Options with which a controller can be configured.
 * The options are used for all instances of a Controller class.
 */
export interface ControllerOptions {
    /**
     * If set to `true` the controller will `throw` whenever an error is encountered.
     * Default is `true`.
     */
    readonly throwOnError?: boolean;
    /**
     * When supplied, this handler will be called for all errors encountered by this controller.
     */
    readonly errorHandler?: ErrorHandler;
    /**
     * A base Url on which the backend can be publicly reached.
     */
    readonly baseUrl?: string;
    /**
     * Can be `ControllerMode.CLIENT` or `ControllerMode.SERVER`. This is autodetected by default
     * and `ControllerMode.CLIENT` will be used in the browser. This determines if the controller
     * will be used to serve the Api or consume it.
     */
    readonly mode?: ControllerMode;
}

export class Controller {
    public throwOnError = true;
    public errorHandler: ErrorHandler;
    public baseUrl: string;
    public mode: ControllerMode = getDefaultControllerMode();

    constructor(options: ControllerOptions) {
        if (options) {
            this.configure(options);
        }
    }

    /**
     * Will be called by `configureController` and applies the given options to this controller.
     */
    public configure(options: ControllerOptions) {
        const { throwOnError, errorHandler, baseUrl, mode } = options;
        if (typeof throwOnError !== "undefined") { this.throwOnError = throwOnError; }
        if (typeof errorHandler !== "undefined") { this.errorHandler = errorHandler; }
        if (typeof baseUrl !== "undefined") { this.baseUrl = baseUrl; }
        if (typeof mode !== "undefined") { this.mode = mode; }
    }

    /**
     * Will be called instead of each route's body when running in `CLIENT` mode. This function
     * will perform the `fetch` against the backend endpoint.
     *
     * @param route The Route meta information for the method for which the functions was called.
     * @param parameters The Url parameters.
     * @param body The body.
     * @param query The query parameters.
     *
     * @return The unwrapped answer from the backend.
     */
    public async wrappedFetch<T>(route: Route, urlParameters: Params, body: any, query: Params): Promise<T> {
        try {
            // Create the url from the given inputs.
            const url = buildUrl(urlParameters, query, this.baseUrl, route.url);

            // Prepare the headers.
            const headers = new Headers();
            headers.append("content-type", "application/json");

            // Perform the actual fetch.
            const response = await fetch(url, {
                body: JSON.stringify(body),
                headers,
                method: route.method,
            });

            // Parse the response from the Api as JSON.
            const answer = await response.json();

            // If the response's status code was a 2xx status code, return the response as it succeeded.
            if (response.ok) {
                const routeData = answer.data;
                if (typeof route.returnType !== "undefined" && typeof route.scope !== "undefined") {
                    if (Array.isArray(routeData)) {
                        return populate(route.scope, Array, route.returnType, routeData);
                    }
                    return populate(route.scope, route.returnType, routeData);
                }
                return routeData;
            }

            // Otherwise, create an error and handle it according to the controller's configuration.
            // This happens by throwing it, which will raise it to the `catch` block below.
            // That block will then perform the actual error handling.
            const error = new ApiError(response.status, answer);
            throw error;
        } catch (error) {
            // Error's might occur as the network might be interrupted or the response might be malformed.
            // Also an error might be thrown by the `try` block above, if a non-2xx status code has been received.
            if (this.errorHandler) {
                this.errorHandler(error);
            }
            if (this.throwOnError) {
                throw error;
            }
        }
    }
}

export function controller(options?: ControllerOptions): ClassDecorator;
export function controller<U, T extends Constructable<U>>(target: T): T;
/**
 * Decorate an individual class as an Api controller.
 *
 * @param options Global options with which this controller can be configured. `configureController` can also set
 *                set and override these options.
 *
 * @return The decorated class.
 */
export function controller<T extends Function>(arg1: ControllerOptions | T): ClassDecorator | T {
    let decorator: (target: T) => T;
    const options = typeof arg1 === "object" ? arg1 : undefined;
    decorator = function(target: T): T {
        Reflect.defineMetadata("api:controller", new Controller(options), target);
        return target;
    };
    if (typeof arg1 === "function") {
        return decorator(arg1 as T);
    }
    return decorator as ClassDecorator;
}
