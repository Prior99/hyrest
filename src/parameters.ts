export interface Params {
    [key: string]: string;
}
