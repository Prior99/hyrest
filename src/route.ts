import "reflect-metadata";

import { HTTPMethod, Params } from "./types";
import { ControllerMode, Controller } from "./controller";
import {
    QueryParameter,
    BodyParameter,
    UrlParameter,
    getBodyParameters,
    getQueryParameters,
    getUrlParameters,
} from "./parameters";

/**
 * Additional options which can be passed to a Route.
 */
export interface RouteOptions {
}

/**
 * A route which can be called via REST.
 */
export interface Route {
    /**
     * The target object which had been decorated with this route.
     * This is the class decorated with @controller.
     */
    readonly target: Object;
    /**
     * The name of the method decorated with @route on `target`.
     */
    readonly property: string;
    /**
     * The HTTP method ("GET", "POST", ...).
     */
    readonly method: HTTPMethod;
    /**
     * The express-compatible Url pattern.
     */
    readonly url: string;
    /**
     * Additional configuration options for this route.
     */
    readonly options?: RouteOptions;
}

/**
 * Returns a list of all routes on `target`. This function is always guaranteed to return an array.
 * If no routes existed on `target` a new refelect metadata key will be created and a new array will
 * be returned.
 *
 * The array can be used to append new routes.
 *
 * @param target The class decorated with @controller for which all routes should be listed.
 *
 * @return An array of all routes on the given `target.
 */
export function getRoutes(target: Object): Route[] {
    const routes = Reflect.getMetadata("api:routes", target);
    if (routes) {
        return routes;
    }
    const newRoutes: Route[] = [];
    Reflect.defineMetadata("api:routes", newRoutes, target);
    return newRoutes;
}

/**
 * This decorated can be used to define a new REST route on a specific controller.
 * Decorate a method with it and the method will be called if a matching request has been caught by the
 * express middleware.
 *
 * **Example:**
 * ```
 * @route("GET", "/user/:id")
 * public getUser(...) {
 * ```
 *
 * @param method The HTTP method ("GET", "POST", ..) to which this route should be sensitive.
 * @param url An express-compatiable Url pattern (e.g. `/user/:id`) which this route should match.
 * @param options Additional options for the route. @see RouteOptions
 *
 * @return A method decorator.
 */
export function route(method: HTTPMethod, url: string, options?: RouteOptions): MethodDecorator {
    return function (target: Object, property: string, descriptor: PropertyDescriptor) {
        // Insert the new `Route` into the reflection metadata.
        const routeMeta = { target, property, method, url, options };
        const routes = getRoutes(target);
        routes.push(routeMeta);

        // Get all injected parameters.
        const queryParameters = getQueryParameters(target, property);
        const bodyParameters = getBodyParameters(target, property);
        const urlParameters = getUrlParameters(target, property);

        // Store the original function before overwriting the method with a wrapper calling the REST endpoint
        // instead if the parent @controller is in `CLIENT` mode.
        const originalFunction = descriptor.value;
        descriptor.value = function (...args: any[]) {
            // Get the parent @controller and throw an error if the route was created in a non-@controller class.
            const controller: Controller = Reflect.getMetadata("api:controller", target.constructor);
            if (!controller) {
                const name = target.constructor.name;
                throw new Error(`Found @route on a class without @controller. Take a look at ${name}.`);
            }

            // If in `CLIENT` mode, call the REST endpoint instead.
            if (controller.mode === ControllerMode.CLIENT) {
                // Find all query parameters based on the @query decorators.
                const query = queryParameters.reduce((result, param) => {
                    result[param.name] = args[param.index];
                    return result;
                }, {} as Params);
                // Find a body paramter based on the @body decorator.
                const body = bodyParameters.length > 0 ? args[bodyParameters[0].index] : undefined;
                // Find all Url paramters based on the @param decorators.
                const parameters = urlParameters.reduce((result, param) => {
                    result[param.name] = args[param.index];
                    return result;
                }, {} as Params);
                // Make the actual REST call and return it's Promise.
                return controller.wrappedFetch(routeMeta, parameters, body, query);
            }

            // Otherwise, just call the original function with the original arguments.
            return originalFunction.apply(this, args); // tslint:disable-line
        };
        return descriptor;
    };
}
