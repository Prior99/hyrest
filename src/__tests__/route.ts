import { route, getRoutes } from "../route";
import { controller, ControllerMode } from "../controller";
import { body, param, query } from "../parameters";
import { ok } from "../answers";

test("`route()` throws when decorating a method on a non-@controller class", () => {
    expect(() => {
        class NotAController {
            @route("GET", "/get")
            public method() {
                return;
            }
        }

        const noController = new NotAController();
        noController.method();
    }).toThrow();
});

test("`route()` registers on a class", () => {
    @controller()
    class TestController {
        @route("GET", "/get")
        public method() {
            return;
        }
    }

    const test = new TestController();
    expect(getRoutes(test)).toMatchSnapshot();
});

test("when calling the route with the controller in `CLIENT` mode", () => {
    @controller({ mode: ControllerMode.CLIENT })
    class TestController {
        @route("GET", "/get/:id")
        public method(@param("id") id: string, @body() thing: any, @query("search") search: string) {
            return;
        }
    }

    const mock = jest.fn();

    const test = new TestController();
    Reflect.getMetadata("api:controller", TestController).wrappedFetch = mock;
    const routes = Reflect.getMetadata("api:routes", test);
    test.method("some-id", { some: "Thing." }, "a query");
    expect(mock).toHaveBeenCalledWith({
        method: "GET",
        url: "/get/:id",
        property: "method",
        target: routes[0].target,
        options: undefined,
    }, {
        id: "some-id",
    }, {
        some: "Thing.",
    }, {
        search: "a query",
    });
});

test("when calling the route with the controller in `CLIENT` mode with no injects", () => {
    @controller({ mode: ControllerMode.CLIENT })
    class TestController {
        @route("GET", "/get/:id")
        public method() {
            return;
        }
    }

    const mock = jest.fn();

    const test = new TestController();
    Reflect.getMetadata("api:controller", TestController).wrappedFetch = mock;
    const routes = Reflect.getMetadata("api:routes", test);
    test.method();
    expect(mock).toHaveBeenCalledWith({
        method: "GET",
        url: "/get/:id",
        property: "method",
        target: routes[0].target,
        options: undefined,
    }, {}, undefined, {});
});

test("when calling the route with the controller in `SERVER` mode", () => {
    @controller({ mode: ControllerMode.SERVER })
    class TestController {
        @route("GET", "/get/:id")
        public method(@param("id") id: string, @body() thing: any, @query("search") search: string) {
            return ok({ id, thing, search });
        }
    }

    const test = new TestController();
    expect(test.method("some-id", { property: "test" }, "query")).toEqual({
        id: "some-id",
        thing: { property: "test" },
        search: "query",
    });
});

test("@route preserves `this`", () => {
    const mock = jest.fn();

    @controller({ mode: ControllerMode.SERVER })
    class TestController {
        @route("GET", "/get")
        public method() {
            mock(this);
            return ok();
        }
    }

    const test = new TestController();
    test.method();
    expect(mock.mock.calls[0][0]).toBe(test);
});
