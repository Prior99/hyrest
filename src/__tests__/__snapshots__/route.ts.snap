// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`route()\` registers on a class 1`] = `
Array [
  Object {
    "method": "GET",
    "options": undefined,
    "property": "method",
    "returnType": undefined,
    "scope": undefined,
    "target": TestController {},
    "url": "/get",
  },
]
`;
